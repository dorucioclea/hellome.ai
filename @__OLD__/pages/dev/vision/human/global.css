@font-face {
  font-style: normal;
  font-weight: 100;
  src: local("Lato"), url("/lato-light.woff2");
  font-family: "Lato";
  font-display: swap;
}
html {
  font-variant: small-caps;
  font-size: 16px;
  font-family: "Lato", "Segoe UI";
}
body {
  margin: 0;
  background: black;
  width: 100vw;
  height: 100vh;
  overflow-x: hidden;
  color: white;
}
body::-webkit-scrollbar {
  display: none;
}
a {
  color: inherit;
  text-decoration: none;
}
