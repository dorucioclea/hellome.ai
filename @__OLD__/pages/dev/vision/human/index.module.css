/* .container {
  padding: 0 2rem;
} */
.footer {
  display: flex;
  justify-content: center;
  align-items: center;
  border-top: 1px solid #eaeaea;
  padding: 1rem;
}
.webcam {
  display: none;
}
.output {
  margin: 0 auto;
  width: 100%;
}
.status {
  position: absolute;
  top: 12px;
  right: 20px;
  box-shadow: 2px 2px black;
  background-color: grey;
  padding: 8px;
  font-family: monospace;
}
.log {
  border-top: 1px solid #eaeaea;
  padding: 8px;
}
.performance {
  position: absolute;
  bottom: 0;
  padding: 8px;
  width: 100%;
  font-size: 0.8rem;
  font-family: monospace;
}
