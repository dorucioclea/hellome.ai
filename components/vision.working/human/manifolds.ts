import type { FaceResult, Result } from "@vladmandic/human"

import {
  distanceMetric,
  withinConfidenceBounds,
  withinToleranceBounds,
} from "@/components/vision.working/human/metrics"

export type UsableFaceResult = FaceResult & {
  rotation: NonNullable<FaceResult["rotation"]>
}

export type Point2D<T = number> = {
  x: T // pitch
  y: T // yaw
  z: T // roll
}

export type ManifoldCoordinates = {
  axis: Point2D<number[]>
  tolerance: Point2D
  confidence: {
    min: {
      face: number
    }
  }
}

export type ManifoldVector = {
  label: string
  index: Point2D
  origin: Point2D
  tolerance: ManifoldCoordinates["tolerance"]
  candidate?: {
    result: Result
    face: FaceResult
    score: number
  }
}

export function castManifoldVectors({
  axis,
  tolerance,
}: ManifoldCoordinates): ManifoldVector[] {
  return axis.x.flatMap((x, xIndex) => {
    return axis.y.flatMap((y, yIndex) => {
      return axis.z.flatMap((z, zIndex) => {
        const vector: ManifoldVector = {
          label: `x:${x * 100}° y:${y * 100}° z:${z * 100}°`,
          index: { x: xIndex, y: yIndex, z: zIndex },
          origin: { x, y, z },
          tolerance,
        }
        return [vector]
      })
    })
  })
}

export function findNearestManifoldVector(
  coordinates: ManifoldCoordinates,
  vectors: ManifoldVector[],
  face: UsableFaceResult,
): ManifoldVector | undefined {
  if (vectors.length === 0 || !withinConfidenceBounds(face, coordinates)) {
    return undefined
  }

  const toleratedVectors = vectors.filter((vector) =>
    withinToleranceBounds(face, vector),
  )

  if (toleratedVectors.length === 0) {
    return undefined
  }

  return toleratedVectors.reduce((prev, current) => {
    return distanceMetric(face, prev) < distanceMetric(face, current)
      ? current
      : prev
  })
}
