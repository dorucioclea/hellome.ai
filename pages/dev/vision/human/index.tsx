import { useRef } from "react"
import Head from "next/head"

import InitWebcam from "@/components/vision/InitWebcam"
import FacePredictionOverlay from "@/components/vision/human/FacePredictionOverlay/FacePredictionOverlay"

import styles from "@/pages/dev/vision/human/index.module.css"

// import { Layout } from '@/components/common'

export default function Index() {
  const canvasRef = useRef<HTMLCanvasElement>(null)
  const videoRef = useRef<HTMLVideoElement>(null)
  const statusRef = useRef<HTMLDivElement>(null)
  const logRef = useRef<HTMLDivElement>(null)
  const performanceRef = useRef<HTMLDivElement>(null)

  return (
    <div className={styles.container}>
      <Head>
        <title>Human</title>
        <meta
          name="description"
          content="Human: Demo with TypeScript/ReactJS/NextJS"
        />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      {/* placeholder element that will be used by human for output */}
      <canvas ref={canvasRef} id="canvas" className={styles.output} />
      <video
        ref={videoRef}
        id="video"
        className={styles.webcam}
        autoPlay
        muted
      />

      {/* <Webcam id="video" className={styles.webcam} autoPlay muted /> */}

      {/* placeholder element that will be used by webcam */}
      <div ref={statusRef} id="status" className={styles.status} />
      <div ref={logRef} id="log" className={styles.log} />
      <div
        ref={performanceRef}
        id="performance"
        className={styles.performance}
      />
      <InitWebcam elementId="video" />

      {/* initialized webcam using html video element with specified id */}
      <FacePredictionOverlay
        inputId="video"
        outputId="canvas"
        canvasRef={canvasRef}
        videoRef={videoRef}
        statusRef={statusRef}
        logRef={logRef}
        performanceRef={performanceRef}

        // handleResult={handleHumanResult}
      />
    </div>
  )
}

// Index.Layout = Layout
