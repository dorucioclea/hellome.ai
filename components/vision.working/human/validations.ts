import { Result } from "@vladmandic/human"

import { ManifoldCoordinates } from "@/components/vision.working/human/manifolds"

export function validateResult(
  coordinates: ManifoldCoordinates,
  result: Result,
): boolean {
  if (result.face.length === 0) {
    // console.warn(`Found no faces`, result)
    return false
  }
  if (result.face.length > 1) {
    console.warn(`Found multiple faces: ${result.face.length}`, result)
    return false
  }
  return true
}

export function isUsable(
  coordinates: ManifoldCoordinates,
  result: Result,
): boolean {
  if (result.face.length === 0) {
    // console.warn(`Found no faces`, result)
    return false
  }
  if (result.face.length > 1) {
    console.warn(`Found multiple faces: ${result.face.length}`, result)
    return false
  }
  if (!result.face[0].rotation) {
    console.warn(`Found no face.rotation`, result)
    return false
  }
  return true
}
