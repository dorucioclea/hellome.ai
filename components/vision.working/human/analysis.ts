import type { Result } from "@vladmandic/human"
import { indexOf } from "lodash"

import {
  ManifoldCoordinates,
  ManifoldVector,
  UsableFaceResult,
  findNearestManifoldVector,
} from "./manifolds"
import { scoreFaceResult } from "./metrics"
import { isUsable } from "./validations"

export type FaceAnalysis = {
  coordinates: ManifoldCoordinates
  vectors: ManifoldVector[]
}

export type FaceAnalysisResult = {
  disposeResults: Result[]
  analysis: FaceAnalysis
}

export function runFaceAnalysis(
  result: Result,
  prevAnalysis: FaceAnalysis,
): FaceAnalysisResult {
  if (!isUsable(prevAnalysis.coordinates, result)) {
    return {
      disposeResults: [result],
      analysis: prevAnalysis,
    }
  }

  // const {coordinates, vectors}= prevAnalysis

  const face = result.face[0] as UsableFaceResult

  const prevVector = findNearestManifoldVector(
    prevAnalysis.coordinates,
    prevAnalysis.vectors,
    face,
  )
  const index = indexOf(prevAnalysis.vectors, prevVector)

  const nextVectors = prevAnalysis.vectors
  if (prevVector && index) {
    const score = scoreFaceResult(face, prevVector)

    if (score > (prevVector.candidate?.score ?? Number.NEGATIVE_INFINITY)) {
      const nextVector = {
        ...prevVector,
        candidate: {
          result,
          face,
          score,
        },
      }

      nextVectors[index] = nextVector
    }

    return {
      disposeResults: prevVector.candidate?.result
        ? [prevVector.candidate.result]
        : [],
      analysis: {
        ...prevAnalysis,
        vectors: nextVectors,
      },
    }
  }

  return {
    disposeResults: [result],
    analysis: prevAnalysis,
  }
}
