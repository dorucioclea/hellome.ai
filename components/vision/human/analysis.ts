import type { FaceResult, Human, Result } from "@vladmandic/human"
import { indexOf, isEqual } from "lodash"

import {
  ManifoldCoordinates,
  ManifoldVector,
  ManifoldVectorCandidate,
  UsableFaceResult,
  findNearestManifoldVector,
} from "./manifolds"
import { scoreFaceResult } from "./metrics"
import { isUsable } from "./validations"

export type FaceAnalysis = {
  coordinates: ManifoldCoordinates
  vectors: ManifoldVector<Result, FaceResult>[]
}

export type FaceAnalysisResult = {
  disposeResults: Result[]
  analysis: FaceAnalysis
}

export async function runFaceAnalysis(
  human: Human | undefined,
  result: Result,
  prevAnalysis: FaceAnalysis,
): Promise<FaceAnalysisResult> {
  if (!isUsable(prevAnalysis.coordinates, result)) {
    return {
      disposeResults: [result],
      analysis: prevAnalysis,
    }
  }

  // const {coordinates, vectors}= prevAnalysis

  const face = result.face[0] as UsableFaceResult

  const prevVector = findNearestManifoldVector(
    prevAnalysis.coordinates,
    prevAnalysis.vectors,
    face,
  )
  // let imageData: ImageData

  const index = indexOf(prevAnalysis.vectors, prevVector)

  const nextVectors = prevAnalysis.vectors

  if (prevVector && index) {
    const score = scoreFaceResult(face, prevVector)

    if (score > (prevVector.candidate?.score ?? Number.NEGATIVE_INFINITY)) {
      // if (!face.tensor?.isDisposed) {
      //   const [_x, _y, width, height] = face.box
      //   const bytes = await human!.tf.browser.toPixels(face.tensor)

      //   // console.log({height, width, bytes, face})

      //   imageData = new ImageData(bytes, width, height)
      // }

      // dispose
      human!.tf.dispose(face.tensor)

      result.face.forEach((faceResult) => {
        if (!isEqual(face, faceResult)) {
          faceResult.tensor?.dispose()
        }
      })

      const candidate: ManifoldVectorCandidate<Result, FaceResult> = {
        // imageData,
        result,
        face,
        score,
      }

      const nextVector = {
        ...prevVector,
        candidate,
      }

      nextVectors[index] = nextVector
    }

    return {
      disposeResults: prevVector.candidate?.result
        ? [prevVector.candidate.result]
        : [],
      analysis: {
        ...prevAnalysis,
        vectors: nextVectors,
      },
    }
  }

  return {
    disposeResults: [result],
    analysis: prevAnalysis,
  }
}
