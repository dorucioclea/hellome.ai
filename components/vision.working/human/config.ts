import { type FaceAnalysis } from "@/components/vision.working/human/analysis"
import {
  castManifoldVectors,
  type ManifoldCoordinates,
} from "@/components/vision.working/human/manifolds"

export const DEFAULT_COORDINATES: ManifoldCoordinates = {
  axis: {
    x: [-0.45, -0.3, -0.15, 0, 0.15, 0.3, 0.45],
    y: [-0.3, -0.15, 0, 0.15, 0.3],
    z: [0],
  },
  tolerance: {
    x: 0.1,
    y: 0.1,
    z: 0.5,
  },
  confidence: {
    min: {
      face: 0.7,
    },
  },
}

export const defaultFaceAnalysis: FaceAnalysis = {
  coordinates: DEFAULT_COORDINATES,
  vectors: castManifoldVectors(DEFAULT_COORDINATES),
}
