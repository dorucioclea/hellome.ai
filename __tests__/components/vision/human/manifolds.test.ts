import { FaceResult, Result } from "@vladmandic/human"

import {
  castManifoldVectors,
  findNearestManifoldVector,
} from "@/components/vision/human/manifolds"

import { DEFAULT_COORDINATES, DEFAULT_USABLE_FACE_RESULT } from "./fixtures"

describe("Given: DEFAULT_COORDINATES", () => {
  const coordinates = DEFAULT_COORDINATES

  describe(".castManifoldVectors/1", () => {
    it("matches snapshot", () => {
      const vectors = castManifoldVectors<Result, FaceResult>(coordinates)
      expect(vectors).toMatchSnapshot()
    })
  })

  describe(".findNearestManifoldVector/3", () => {
    const face = DEFAULT_USABLE_FACE_RESULT
    const vectors = castManifoldVectors<Result, FaceResult>(coordinates)

    it("DEFAULT_USABLE_FACE_RESULT matches vectors[0]", () => {
      const result = findNearestManifoldVector(coordinates, vectors, face)
      expect(result).toEqual(vectors[0])
    })

    it("handles empty vectors", () => {
      const result = findNearestManifoldVector(coordinates, [], face)
      expect(result).toEqual(undefined)
    })
  })
})
