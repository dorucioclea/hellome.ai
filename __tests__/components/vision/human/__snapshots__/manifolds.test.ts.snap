// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given: DEFAULT_COORDINATES .castManifoldVectors/1 matches snapshot 1`] = `
[
  {
    "index": {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "label": "x:-45° y:-30° z:0°",
    "origin": {
      "x": -0.45,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 0,
      "y": 1,
      "z": 0,
    },
    "label": "x:-45° y:-15° z:0°",
    "origin": {
      "x": -0.45,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 0,
      "y": 2,
      "z": 0,
    },
    "label": "x:-45° y:0° z:0°",
    "origin": {
      "x": -0.45,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 0,
      "y": 3,
      "z": 0,
    },
    "label": "x:-45° y:15° z:0°",
    "origin": {
      "x": -0.45,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 0,
      "y": 4,
      "z": 0,
    },
    "label": "x:-45° y:30° z:0°",
    "origin": {
      "x": -0.45,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 1,
      "y": 0,
      "z": 0,
    },
    "label": "x:-30° y:-30° z:0°",
    "origin": {
      "x": -0.3,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 1,
      "y": 1,
      "z": 0,
    },
    "label": "x:-30° y:-15° z:0°",
    "origin": {
      "x": -0.3,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 1,
      "y": 2,
      "z": 0,
    },
    "label": "x:-30° y:0° z:0°",
    "origin": {
      "x": -0.3,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 1,
      "y": 3,
      "z": 0,
    },
    "label": "x:-30° y:15° z:0°",
    "origin": {
      "x": -0.3,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 1,
      "y": 4,
      "z": 0,
    },
    "label": "x:-30° y:30° z:0°",
    "origin": {
      "x": -0.3,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 2,
      "y": 0,
      "z": 0,
    },
    "label": "x:-15° y:-30° z:0°",
    "origin": {
      "x": -0.15,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 2,
      "y": 1,
      "z": 0,
    },
    "label": "x:-15° y:-15° z:0°",
    "origin": {
      "x": -0.15,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 2,
      "y": 2,
      "z": 0,
    },
    "label": "x:-15° y:0° z:0°",
    "origin": {
      "x": -0.15,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 2,
      "y": 3,
      "z": 0,
    },
    "label": "x:-15° y:15° z:0°",
    "origin": {
      "x": -0.15,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 2,
      "y": 4,
      "z": 0,
    },
    "label": "x:-15° y:30° z:0°",
    "origin": {
      "x": -0.15,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 3,
      "y": 0,
      "z": 0,
    },
    "label": "x:0° y:-30° z:0°",
    "origin": {
      "x": 0,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 3,
      "y": 1,
      "z": 0,
    },
    "label": "x:0° y:-15° z:0°",
    "origin": {
      "x": 0,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 3,
      "y": 2,
      "z": 0,
    },
    "label": "x:0° y:0° z:0°",
    "origin": {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 3,
      "y": 3,
      "z": 0,
    },
    "label": "x:0° y:15° z:0°",
    "origin": {
      "x": 0,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 3,
      "y": 4,
      "z": 0,
    },
    "label": "x:0° y:30° z:0°",
    "origin": {
      "x": 0,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 4,
      "y": 0,
      "z": 0,
    },
    "label": "x:15° y:-30° z:0°",
    "origin": {
      "x": 0.15,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 4,
      "y": 1,
      "z": 0,
    },
    "label": "x:15° y:-15° z:0°",
    "origin": {
      "x": 0.15,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 4,
      "y": 2,
      "z": 0,
    },
    "label": "x:15° y:0° z:0°",
    "origin": {
      "x": 0.15,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 4,
      "y": 3,
      "z": 0,
    },
    "label": "x:15° y:15° z:0°",
    "origin": {
      "x": 0.15,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 4,
      "y": 4,
      "z": 0,
    },
    "label": "x:15° y:30° z:0°",
    "origin": {
      "x": 0.15,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 5,
      "y": 0,
      "z": 0,
    },
    "label": "x:30° y:-30° z:0°",
    "origin": {
      "x": 0.3,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 5,
      "y": 1,
      "z": 0,
    },
    "label": "x:30° y:-15° z:0°",
    "origin": {
      "x": 0.3,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 5,
      "y": 2,
      "z": 0,
    },
    "label": "x:30° y:0° z:0°",
    "origin": {
      "x": 0.3,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 5,
      "y": 3,
      "z": 0,
    },
    "label": "x:30° y:15° z:0°",
    "origin": {
      "x": 0.3,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 5,
      "y": 4,
      "z": 0,
    },
    "label": "x:30° y:30° z:0°",
    "origin": {
      "x": 0.3,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 6,
      "y": 0,
      "z": 0,
    },
    "label": "x:45° y:-30° z:0°",
    "origin": {
      "x": 0.45,
      "y": -0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 6,
      "y": 1,
      "z": 0,
    },
    "label": "x:45° y:-15° z:0°",
    "origin": {
      "x": 0.45,
      "y": -0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 6,
      "y": 2,
      "z": 0,
    },
    "label": "x:45° y:0° z:0°",
    "origin": {
      "x": 0.45,
      "y": 0,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 6,
      "y": 3,
      "z": 0,
    },
    "label": "x:45° y:15° z:0°",
    "origin": {
      "x": 0.45,
      "y": 0.15,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
  {
    "index": {
      "x": 6,
      "y": 4,
      "z": 0,
    },
    "label": "x:45° y:30° z:0°",
    "origin": {
      "x": 0.45,
      "y": 0.3,
      "z": 0,
    },
    "tolerance": {
      "x": 0.1,
      "y": 0.1,
      "z": 0.5,
    },
  },
]
`;
